src/number_theory/modular_forms/jacobi_theta.lean

```diff
@@ -0,0 +1,55 @@
+/-
+Copyright (c) 2023 David Loeffler. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Loeffler
+-/
+
+import analysis.fourier.poisson_summation
+import analysis.schwartz_space
+import analysis.p_series
+import analysis.special_functions.gaussian
+
+/-! # Jacobi's theta function
+
+This file defines the Jacobi theta function
+
+$$\theta(\tau) = \sum_{n \in \mathbb{Z}} \exp (i \pi n ^ 2 \tau),$$
+
+and proves the modular transformation properties `θ (τ + 2) = θ τ` and
+`θ (-1 / τ) = (-I * τ) ^ (1 / 2) * θ τ`, using Poisson's summation formula for the latter.
+-/
+
+open filter asymptotics function set complex (hiding abs_of_nonneg sq_abs) real (hiding exp)
+
+open_locale real topology big_operators filter fourier_transform
+
+/-- Jacobi's theta function `∑' (n : ℤ), exp (π * I * n ^ 2 * τ)`. -/
+noncomputable def jacobi_theta (τ : ℂ) : ℂ := ∑' (n : ℤ), cexp (π * I * n ^ 2 * τ)
+
+lemma jacobi_theta_add_two : periodic jacobi_theta 2 :=
+begin
+  refine λ τ, tsum_congr (λ n, _),
+  suffices : cexp (↑π * I * ↑n ^ 2 * 2) = 1, by rw [mul_add, complex.exp_add, this, mul_one],
+  rw [(by { push_cast, ring } : ↑π * I * ↑n ^ 2 * 2 = ↑(n ^ 2) * (2 * π * I)),
+    complex.exp_int_mul, complex.exp_two_pi_mul_I, one_zpow],
+end
+
+lemma jacobi_theta_neg_one_div {τ : ℂ} (hτ : 0 < τ.im) :
```

Comment:
Instead of  `{τ : ℂ} (hτ : 0 < τ.im) ` can you just use `{τ : ℍ} `? You then for example get `hτ` from `upper_half_plane.ne_zero`

Reply:
Thanks, that's a good point. I've rewritten it in terms of `τ : ℍ` and the action of `S` and `T ^ 2`.